### ISSUE TITLE
clarify significance of argument ordering to Space.add_collision_handler

### ISSUE THREAD

**Issue by numberoverzero** [viblo/pymunk#132]
This may be a lack of grokking Chipmunk, but I noticed that calling `Space.add_collision_handler` with `*(a, b)` and `*(b, a)` gave different handlers, but only one's `separate` handler was ever called.

Is the ordering significant?  I tried swapping the order that I added collision handlers, but it always picked the same one (ie. not just the last one defined).

This isn't a complete sample, but for the sake of discussion:

```python
import pymunk as pm
space = pm.space()
wall_type = 1
player_type = 2

def wall_player_separate(*args):
    print("wall_player")


def player_wall_separate(*args):
    print("player_wall")


wall_handler = space.add_collision_handler(wall_type, player_type)  # [0]
player_handler = space.add_collision_handler(player_type, wall_type)  # [1]

assert wall_handler is not player_handler  # [2]

wall_handler.separate = wall_player_separate
player_handler.separate = player_wall_separate

...

for i in range(100):
    space.update(1/60)
```

I expected that, since the two handlers are not the same (see [2]) exactly one of the following would happen:

1. They would both be invoked
2. Only the last defined would be invoked
3. Only the first defined would be invoked
4. The second `add_collision_handler` call would return the existing handler, and the assert would fail

Instead, `wall_handler` was always used, no matter the order I defined them in; when I swapped [0] and [1] it still used the wall handler; only when I didn't attach the wall_handler did it use the `player_handler`.

Is this expected?  If only one is going to be used and it's not consistently the first or the second, I'd prefer to at least get the original object back, or throw an exception that a handler for the given pair of collision types is already defined.

**Comment by viblo** [comment]
Thanks for the report! 

When I test this myself it seems like the last separate function to be assigned is always used. 

```
import pymunk as p

s = p.Space()
s.gravity = 0,-1000

b1 = p.Body(body_type=p.Body.STATIC)
c1 = p.Circle(b1, 3)
c1.collision_type = 1
c1.name = "c1"

b2 = p.Body(1, 1)
b2.position = 1,10
c2 = p.Circle(b2, 3)
c2.collision_type = 2
c2.name = "c2"
s.add(b1, c1, b2, c2)

def c1f(a, *args):
    print "c1f", [s.name for s in a.shapes]

def c2f(a, *args):
    print "c2f", [s.name for s in a.shapes]

h2 = s.add_collision_handler(2,1)
h1 = s.add_collision_handler(1,2)

h2.separate = c2f  # B
h1.separate = c1f  # A

for x in range(10):
    s.step(0.1)

```
If I change place of A and B c2f will be invoked instead of c1f. 

But maybe it would be better to do either your [1] or [4]. Doing [1] is a bit complicated, so that leaves [4] as the best option.

**Comment by numberoverzero** [comment]
I like [4] the best; you could canonicalize the two collision types inside `add_collision_handler`:

```python
def add_collision_handler(self, collision_type_a, collision_type_b):
    key = min(collision_type_a, collision_type_b), max(collision_type_a, collision_type_b)
    ...
```

This is still a breaking change for anyone that relies on very specific edge-case behavior: specifically, the order in which callbacks are attached to two different-but-actually-equal CollisionHandler objects.

Another solution I think, is to make `CollisionHandler.__hash__` return `hash(self._handler)` since those should be the same, right?  (Assuming the Chipmunk code de-dupes and is returning the same handler that you wrap)

**Comment by viblo** [comment]
Agreed, I think the min/max is easy to implement and fixes the problem nicely.
Its a bit sad it wont be fully backwards compatible, maybe it needs to wait for 6.0..

**Comment by viblo** [comment]
I have just released Pymunk 6.0.0. This version contains a fix for this issue. Would be great if you had the opportunity to test it! (Note however that this version contains several api breaking changes, so if you have existing code you will very likely need to do some adjustments)
